File: frontend/src/api/versions.js
import { api } from './client'
import { downloadFromApi } from '../lib/download'

export const versionsApi = {
  list: (projectId) => api.get(`/projects/${projectId}/versions`).then(r => r.data),
  get: (id) => api.get(`/versions/${id}`).then(r => r.data),
  create: (projectId, data) => api.post(`/projects/${projectId}/versions`, data).then(r => r.data),
  update: (id, data) => api.put(`/versions/${id}`, data).then(r => r.data),
  delete: (id) => api.delete(`/versions/${id}`),

  listAttachments: (versionId) =>
    api.get(`/versions/${versionId}/attachments`).then(r => r.data),
  uploadAttachment: (versionId, file) => {
    const form = new FormData()
    form.append('file', file)
    return api.post(`/versions/${versionId}/attachments`, form).then(r => r.data)
  },
  deleteAttachment: (versionId, attachmentId) =>
    api.delete(`/versions/${versionId}/attachments/${attachmentId}`),
  downloadAttachment: (versionId, attachmentId, filename) =>
    downloadFromApi(`/versions/${versionId}/attachments/${attachmentId}/download`, filename),
}

File: client/testjam_client/resources/versions.py
"""Project versions endpoints."""
from __future__ import annotations

from typing import Any

from testjam_client.errors import Conflict
from testjam_client.resources._base import Resource


class VersionsResource(Resource):
    def list(self, project_id: int) -> list[dict]:
        return self._request("GET", f"/projects/{project_id}/versions").json()

    def get(self, version_id: int) -> dict:
        return self._request("GET", f"/versions/{version_id}").json()

    def create(
        self,
        project_id: int,
        name: str,
        *,
        tag: str | None = None,
        description: str | None = None,
        release_date: str | None = None,
    ) -> dict:
        body: dict[str, Any] = {"name": name}
        if tag is not None:
            body["tag"] = tag
        if description is not None:
            body["description"] = description
        if release_date is not None:
            body["release_date"] = release_date
        return self._request("POST", f"/projects/{project_id}/versions", json=body).json()

    def update(self, version_id: int, **payload: Any) -> dict:
        return self._request("PUT", f"/versions/{version_id}", json=payload).json()

    def delete(self, version_id: int) -> None:
        self._request("DELETE", f"/versions/{version_id}")

    def find_by_name(self, project_id: int, name: str) -> dict | None:
        for version in self.list(project_id):
            if version["name"] == name:
                return version
        return None

    def find_or_create(
        self,
        project_id: int,
        name: str,
        **kwargs: Any,
    ) -> dict:
        existing = self.find_by_name(project_id, name)
        if existing is not None:
            return existing
        try:
            return self.create(project_id, name, **kwargs)
        except Conflict:
            existing = self.find_by_name(project_id, name)
            if existing is None:
                raise
            return existing

    def list_attachments(self, version_id: int) -> list[dict]:
        return self._request("GET", f"/versions/{version_id}/attachments").json()

    def upload_attachment(
        self,
        version_id: int,
        *,
        filename: str,
        content,
        mime: str,
    ) -> dict:
        files = {"file": (filename, content, mime)}
        return self._request(
            "POST", f"/versions/{version_id}/attachments", files=files,
        ).json()

    def download_attachment(self, version_id: int, attachment_id: int) -> bytes:
        response = self._request(
            "GET", f"/versions/{version_id}/attachments/{attachment_id}/download",
        )
        return response.content

    def delete_attachment(self, version_id: int, attachment_id: int) -> None:
        self._request(
            "DELETE", f"/versions/{version_id}/attachments/{attachment_id}",
        )
